README_B2C.md

```diff
@@ -92,27 +84,26 @@ In the steps below, "ClientID" is the same as "Application ID" or "AppId".
 
    * Update the value of `b2c_tenant` with the name of the Azure AD B2C tenant that you created.
      For example, replace `fabrikamb2c` with `contoso`.
-   * Replace the value of `CLIENT_ID` with the application ID that you recorded.
-   * Replace the value of `CLIENT_SECRET` with the key that you recorded.
-   * Replace the value of `signupsignin_user_flow` with `b2c_1_signupsignin1`.
-   * Replace the value of `editprofile_user_flow` with `b2c_1_profileediting1`.
-   * Replace the value of `resetpassword_user_flow` with `b2c_1_passwordreset1`.
-   * Replace the value of `REDIRECT_PATH` with the path part you set up in **Reply URL**.
+   * Replace the value of `CLIENT_ID` with the Application (client) ID that you recorded.
+   * Replace the value of `CLIENT_SECRET` with the client secret that you recorded.
+   * Replace the value of `signupsignin_user_flow` with `B2C_1_signupsignin1`.
+   * Replace the value of `editprofile_user_flow` with `B2C_1_profileediting1`.
+   * Replace the value of `resetpassword_user_flow` with `B2C_1_passwordreset1`.
+   * Replace the value of `REDIRECT_PATH` with the path part you set up in **Redirect URIs**.
      For example, `/getAToken`. It will be used by this sample app to form
-     an absolute URL which matches your full **Reply URL**.
+     an absolute URL which matches your full **Redirect URI**.
    * You do not have to configure the `ENDPOINT` and `SCOPE` right now
 
-
 ## Enable authentication
 
-Run app.py from shell or command line. Note that the port needs to match what you've set up in your **Reply URL**:
+Run app.py from shell or command line. Note that the port needs to match what you've set up in your **Redirect URI**:
+
 ```Shell
-$ flask run --port 5000
+$ flask run --host localhost --port 5000
```

Comment:
It seems not strictly required. I am fine with either way. But then if you have reason to change it to this way, you might as well also modify its counterpart in README.md :-)

Reply:
@rayluo Ah, yes, here's why I went with that. Without specifying host arg, the URL the server spits out is `http://127.0.0.1:5000/`. If you navigate to the URL it specifies and select the **Sign In** link, your redirect URI is mismatched and you get a `redirect_uri_mismatch`.

Here's without the `--host` arg:

```console
(flask-01) marsh@marsh:~/repos/ms-identity-python-webapp$ flask run --port 5000
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [11/Jun/2020 13:24:00] "GET / HTTP/1.1" 302 -
127.0.0.1 - - [11/Jun/2020 13:24:00] "GET /login HTTP/1.1" 200 -
127.0.0.1 - - [11/Jun/2020 13:24:02] "GET /getAToken?error=redirect_uri_mismatch&error_description=AADB2C90006%3a+The+redirect+URI+%27http%3a%2f%2f127.0.0.1%3a5000%2fgetAToken%27+provided+in+the+request+is+not+registered+for+the+client+id+%27c52ee87d-666f-4164-94f2-bdcac14f7e34%27.%0d%0aCorrelation+ID%3a+76225db9-039e-483c-a8b9-444b91207b5d%0d%0aTimestamp%3a+2020-06-11+20%3a24%3a02Z%0d%0a&state=172a492d-ceb8-45ca-b2bf-fdbdb8621be6 HTTP/1.1" 302 -
127.0.0.1 - - [11/Jun/2020 13:24:02] "GET / HTTP/1.1" 302 -
127.0.0.1 - - [11/Jun/2020 13:24:03] "GET /login HTTP/1.1" 200 -
``` 

And with the `--host` arg:

```console
^C(flask-01) marsh@marsh:~/repos/ms-identity-python-webapp$ flask run --host localhost --port 5000
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://localhost:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [11/Jun/2020 13:29:49] "GET / HTTP/1.1" 302 -
127.0.0.1 - - [11/Jun/2020 13:29:50] "GET /login HTTP/1.1" 200 -
127.0.0.1 - - [11/Jun/2020 13:29:59] "GET /getAToken?state=90bae38b-5768-4c5f-814c-a0bda0eb182c&code=eyJraWQiOiJjcGltY29yZV8redactedIjEuMCIsInppcCI6IkRlZmredactediOiIxLjAifQ..yd2FredactedBxL.8xVNFJP9pMrZ8redacted_uBcYhgjhcyKZ5tRvaGredacted9GFmo-RMredactedt8ePfDui9vAQQQVnRk_EH9jDT0jzGMvpHGSredactedWhntTm-P0i57gredactedqBNuzOglHvpsP5DEGs_v9A3C0WMGredactedu6FPCEW5o6xredactedv1h9tPoplg68yw96-dZ_sPiNztKnQuOGmTjFmCN6aa0EamhoTte-yArcYvb-jDbUCc_M0dredactedXlBitX4kdrYFTQontQMB1huBFZIUvp0b9vWG20V2PMwl3pULxB32yQVj26FE0tgFredactedeiLGhcBrQqdSUlj8g_tVzqMtXfTI4ElhEfJr2zredactedFYnen2FAdOulf5lz3n_hJjKWPCGredactedeuUZeyojuaugeHzcOYzmIaWnaXvredactedNSpNliM1tq8_06redactedjIb58QaFcVhWz7UZUQvsuGqYxbluh0_bAB2vbY1EckzRWGjvhsCcvEv7PRo2C1mSK8WJX_Bo5CKXMcO2FARweE7-aQiCwredactedM5QdEY228g.FR-kDBfWkM_i70HdXAwlNQ HTTP/1.1" 302 -
127.0.0.1 - - [11/Jun/2020 13:29:59] "GET / HTTP/1.1" 200 -
```

This changes avoids that situation by ensuring that it always launches on the URL for which you've added a redirect URI.

Will update in the main README.md, too! :)

Reply:
Good to know! Admittedly, I did not pay attention to this subtle behavior that AAD treat `localhost` and `127.0.0.1` differently. But that is not really an issue there.

Further test indicates that (and please help re-test it in your environment to double check), there is no difference when running our sample app with `flask run --port 5000` and `flask run --host localhost --port 5000`. Either way, you will get a web server up and running, accepting requests towards EITHER `http://localhost:5000` OR `http://127.0.0.1:5000`. So, `localhost` and `127.0.0.1` are really interchangeable, from a network routing perspective.

The real requirement is, the app developer would have to visit his/her local sample app using either `localhost` or `127.0.0.1`, as long as it matches what has been setup in AAD Portal. And, given that our portal does not accept `http://127.0.0.1` as redirect_uri setup, our README.md should have called it out to ask app developer to call this app using `http://localhost:5000`.

I can send out a minor patch for this repo to automatically redirect app developer from `127.0.0.1` to `localhost`. But our other web samples might not have such behavior, thus it is still conceptually necessarily to call it out in README.md.

So, perhaps you would want to refine README.md and README_B2C.md, one more time?

Reply:
@rayluo Yes, from a local routing perspective, it doesn't matter whether it's the loopback IP or localhost--it's all about matching your redirect URI.

I've already updated the line just above that to also specify that the *host* value must match your redirect URI; previously it mentioned only the port:

![image](https://user-images.githubusercontent.com/13394541/84449010-1118bc80-ac01-11ea-95a7-e18b3db286f9.png)

For now, I think that's sufficient to get the point across. We can look at clarifying if people raise issues, but I don't think additional code or text will make it clearer or be more helpful. In fact, I'm afraid it might even muddle the issue and make it more complex than it needs to be.

Reply:
Well then, I did not realize that your "host ... need to match" wording was so correct, in such a sneaky way. :-)

> I don't think additional code or text will make it clearer or be more helpful. In fact, I'm afraid it might even muddle the issue and make it more complex than it needs to be.

So you mean we do not even need to ship with [this improvement in the sample](https://github.com/Azure-Samples/ms-identity-python-webapp/compare/redirect-loopback-ip-to-localhost?expand=1)? That's also fine. I'll just leave its link here for future reference.

Reply:
@rayluo Yep, I think we can leave it out for now. I'm a huge fan of "clear over clever" in code samples, and this little bit of helper magic could potentially cause confusion if a dev didn't notice it. Sometimes less help is better in order to not "hide" a failure case.